smarts/core/chassis.py

```diff
@@ -629,20 +651,30 @@ def control(self, throttle=0, brake=0, steering=0):
         assert 0 <= brake <= 1, f"brake ({brake}) must be in [0, 1]"
         assert -1 <= steering <= 1, f"steering ({steering}) must be in [-1, 1]"
 
+        # If we apply brake at low speed using reverse torque
+        # the vehicle starts to roll back. we need to apply a condition
+        # on brake such that, the reverse torque is only applied after
+        # a threshold is passed for vehicle velocity.
+        # Thus, brake is applied if: vehicle speed > 1/36 (m/s)
+        if brake > 0 and self.longitudinal_lateral_speed[0] < 1 / 36:
+            brake = 0
+
         self._apply_steering(steering)
+
         # If the tire parameters yaml file exists, then the throttle and
         # brake forces are applied according to the requested tire model.
         # Otherwise, it uses bullet to calculate the reaction forces.
-
         if self._tire_model != None:
             self._lat_forces, self._lon_forces = self._tire_model.apply_tire_forces(
                 self,
                 self.bullet_client,
                 [(1 / self._max_torque) * np.array(throttle_list), brake, steering],
             )
+            self._clear_step_cache()
             return
         self._apply_throttle(throttle_list)
         self._apply_brake(brake)
+        self._clear_step_cache()
```

Comment:
Is it necessary to invalidate the cache when applying immediate forces? These forces do not have effect until after a physics step.

Reply:
Yes, I think so.  I followed the principle here of invalidating the cache whenever a mutating pybullet API call is made (or in this case, doing it only once after the 3 were made above to steering, throttle and brake).   The reason is because at this point, it's not always clear from the pybullet documentation whether subsequent API calls to pybullet will return the same value or not -- i.e., pybullet's documentation doesn't make it clear _which_ of its API calls are only affected by stepping and which ones "take immediate effect".  In this uncertainty, it seems necessary to err on the side of caution.

The cache is still useful though for "downstream" accesses to things like `pose` that happen frequently on a single step from multiple points in SMARTS and shouldn't require repeated pybullet API calls.

Reply:
lol, I don't know what the eyes emoji means!  (feeling old)

Reply:
I am not sure either, I could not find an 👌 so I just used eyes.